crates/wasi-common/src/ctx.rs

```diff
@@ -408,42 +410,52 @@ impl WasiCtx {
     }
 
     /// Check if `WasiCtx` contains the specified raw WASI `fd`.
-    pub(crate) unsafe fn contains_entry(&self, fd: wasi::__wasi_fd_t) -> bool {
-        self.entries.contains_key(&fd)
+    pub(crate) unsafe fn contains_entry(&self, fd: types::Fd) -> bool {
+        self.entries.borrow().contains_key(&fd)
     }
 
     /// Get an immutable `Entry` corresponding to the specified raw WASI `fd`.
-    pub(crate) unsafe fn get_entry(&self, fd: wasi::__wasi_fd_t) -> WasiResult<&Entry> {
-        self.entries.get(&fd).ok_or(WasiError::EBADF)
+    pub(crate) unsafe fn get_entry(&self, fd: types::Fd) -> Result<Ref<Entry>> {
+        if !self.contains_entry(fd) {
+            return Err(Errno::Badf);
+        }
+        Ok(Ref::map(self.entries.borrow(), |entries| {
+            entries.get(&fd).unwrap()
+        }))
     }
 
     /// Get a mutable `Entry` corresponding to the specified raw WASI `fd`.
-    pub(crate) unsafe fn get_entry_mut(&mut self, fd: wasi::__wasi_fd_t) -> WasiResult<&mut Entry> {
-        self.entries.get_mut(&fd).ok_or(WasiError::EBADF)
+    pub(crate) unsafe fn get_entry_mut(&self, fd: types::Fd) -> Result<RefMut<Entry>> {
```

Comment:
One of the downsides of this returning `RefMut` directly is that it runs the risk of making it easy to panic in wasi-common code by accident I think. For example if while you have a `RefMut` you call back into wasm any reentrancy back into wasi would cause a panic due to a failed borrow.

It may just be worth documenting that for now, but wanted to point out that this may be a panic waiting to happen in the future.

Reply:
Also, I don't think this function needs to be `unsafe`?

Reply:
The `unsafe` is there for the raw file descriptor. This is a convention we picked up from `std::fs::File::from_raw_fd`, which isn't necessarily memory-unsafe, but extends the concept of safety to include I/O safety. Or an alternative interpretation is that file descriptors are pointers into a different address space (they can dangle, alias, be out of bounds, be bogus, etc.), and memory safety rules should apply to that address space too.

That said, I think we're still working out exactly how we want to approach this. See https://github.com/bytecodealliance/wasmtime/pull/1202/files#r395300196 for an example.

Reply:
Cool, I'll leave a comment in for now, and we can figure out a fix later. As far as `unsafe` is concerned, as @sunfishcode pointed out, this is "prior art" and was there to signal the unsafety of using I/O. FWIW I'd leave that in for now and we can figure out how to deal with all the unsafety in a future PR?

Reply:
BTW @alexcrichton, lemme know what's your plan for this, i.e., what to use instead of `RefMut`! :-)

Reply:
Hm ok so I would strongly recommend removing `unsafe` here. There's already enough `unsafe` as-is with dealing with raw wasm memory, and this feels like it's overkill.

The unsafety around libstd's I/O was specifically about providing a guarantee that each entity holding a file descriptor owns that file descriptor. That allows safe abstractions around mmap and stuff (in theory). I really think we should not be adding `unsafe` here lightly, and we may just not have that guarantee for wasi-common.

The unsafety here also doesn't really seem to stop anything in practice, every callsite just wraps a `unsafe { ... }` block which just makes it more verbose to call these methods rather than providing the "oh I should think about this" speedbump it seems.

Reply:
Oh I should also mention, I think the only real alternative to returning `RefMut<Entry>` is to return `Rc<Entry>`. That will likely be required anyway if we want to make this multithreaded at some point (probably don't want to hold locks for entireties of syscalls). I don't know of a great way to otherwise handle this. For now it seems fine to document how you need to be careful about holding a borrow for too long.

Reply:
Hmm, right, I think I get it. `Rc` should still allow for obtaining `&mut` of the wrapped type provided there is only one at play at any one time, right? And then in the multithreaded case, we'd just swap it for `Arc`?

Reply:
I don't believe that `&mut` is needed anywhere actually? For example all methods on I/O files are implemented for `&T` so you have to do something like `(&entry.file).read(...)` which is a bit wonky, but there shouldn't be a fundamental need to ever mutate an `Entry`.

If there is though then a `RefCell` can always be used, and the borrows would need to be scoped in duration.

Reply:
You're absolutely right, we could easily operate on refs since we have impl on `&T` for I/O in Rust. That's cool. I'll make this change here already to clean up the code as much as possible (i.e., removing the `mut`s). In terms of `Rc`, I'm guessing we'd still want to wrap the map holding the `Entry`s in an `Rc`?

Reply:
Right, we do since we still need to able to mutate `EntryTable` and *not* `Entry`!

Reply:
The map will want to be in a `RefCell` for sure (was `Rc` a typo?) because we're modifying it behind a shared reference, and then each entry would be in its own `Rc` to get handed out without holding a long-lived borrow.

Reply:
Sorry, yeah, meant `RefCell`.

Right, makes sense!

Reply:
I was wrong actually. We do need to be able to mutate `Entry` in two places: `fd_fdstat_set_rights` and `fd_fdstat_set_flags`.

Reply:
@sunfishcode would you agree at removing the `unsafe` here?

Reply:
Yes. I've been exploring the "pass Fd by reference" idea more, and I'm sufficiently confident that that will provide a way for `get_entry` and `get_entry_mut` to be safe functions, even considering I/O safety, so I think we can leave the unsafe off for now because that's where they'll end up anyway.

Reply:
@alexcrichton Would you be OK if we parked the `&` vs `&mut` cleanup in `wasi-common`? Changing to `&` in most cases while absolutely correct will require some more work with `virtfs` module (at quick glance, not everything is behind a `RefCell` or alike to be able to mutate the interior behind ` &`.

Reply:
Oh sorry yeah, I think cleaning up later is fine, but it'd be nice to remove the `unsafe` before landing

Reply:
Absolutely, I'm on it 👍